### PR TITLE
Add Heap Sensors - free/max block/fragmentation #1578

### PR DESCRIPTION
**Pull request by micronen**
## Description:

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome](https://github.com/esphome/esphome) with YAML changes (if applicable):** esphome/esphome#1578

## Checklist:

  - [X] Branch: `next` is for changes and new documentation that will go public with the next ESPHome release. Fixes, changes and adjustments for the current release should be created against `current`.
  - [ ] Link added in `/index.rst` when creating new documents for new components or cookbook.

### CODE REVIEW

File: components/debug.rst [hunk 1]
```diff
@@ -17,16 +19,61 @@ a bunch of useful information like reset reason, free heap size, ESPHome version
...
+    - Flash id
+    - SDK, Core & Boot versions
+    - Reset reason & information
+    
+- **free** (*Optional*): reports the free heap size in bytes.
+- **fragmentation** (*Optional*): reports the fragmentation metric 
+  (0% is clean, more than ~50% is not harmless)
+- **block** (*Optional*): reports the largest contiguous free RAM block in the heap in bytes, 
+  useful for checking heap fragmentation. 
+- **loop-time** (*Optional*): reports the time between successive loop calls.
```
Comment (oxan) [inline components/debug.rst]:
```suggestion
- **loop_time** (*Optional*): reports the time between successive loop calls.
```

File: components/debug.rst [hunk 2]
```diff
@@ -17,16 +19,61 @@ a bunch of useful information like reset reason, free heap size, ESPHome version
...
+
+- All other options from :ref:`Sensor <config-sensor>`.
```
Comment (oxan) [inline components/debug.rst]: This is not correct, each individual sensor accepts these options, but the component itself does not.

File: components/debug.rst [hunk 3]
```diff
@@ -17,16 +19,61 @@ a bunch of useful information like reset reason, free heap size, ESPHome version
...
+
+Requires ``arduino_version: 2.5.2`` or above see :ref:`esphome-arduino_version`
```
Comment (oxan) [inline components/debug.rst]: This should be noted under the sensors for which it applies, not globally.

File: components/debug.rst [hunk 4]
```diff
@@ -17,16 +19,61 @@ a bunch of useful information like reset reason, free heap size, ESPHome version
...
 
 See Also
 --------
 
+- https://arduino-esp8266.readthedocs.io/en/latest/libraries.html
```
Comment (oxan) [inline components/debug.rst]:
```suggestion
- https://arduino-esp8266.readthedocs.io/en/latest/libraries.html#esp-specific-apis
```

**Review by oxan — CHANGES_REQUESTED** [review]

**Comment by micronen** [comment]
> There hasn't been any activity on this pull request recently. This pull request has been automatically marked as stale because of that and will be closed if no further activity occurs within 7 days. Thank you for your contributions.

This is part of an open PR, which is waiting for a merge.

**Review by jesserockz — CHANGES_REQUESTED** [review]
As this is a feature matched with a PR in `esphome/esphome`, please target your PR to the `next` branch and rebase.

Thanks
Jesse

**Comment by jesserockz** [comment]
@micronen See my above comment please.

**Comment by micronen** [comment]
@jesserockz
The docs are rebased and target to next. 
Do you want me to add anything to the docs?

**Comment by jesserockz** [comment]
There are requested changes above from Oxan still pending.

File: components/debug.rst (hunk 2 above)
Comment (micronen) [reply components/debug.rst]: fixed

File: components/debug.rst (hunk 3 above)
Comment (micronen) [reply components/debug.rst]: fixex

**Comment by micronen** [comment]
@jesserockzI fixed Oxan's comments.

**Pull request merged**